tensordict/tensordict.py

```diff
@@ -2086,45 +2087,65 @@ def __getitem__(self, idx: INDEX_TYPING) -> TensorDictBase:
     __getitems__ = __getitem__
 
     def __setitem__(
-        self, index: INDEX_TYPING, value: Union[TensorDictBase, dict]
+        self,
+        index: INDEX_TYPING,
+        value: Union[TensorDictBase, dict, numbers.Number, COMPATIBLE_TYPES],
     ) -> None:
-        if index is Ellipsis or (isinstance(index, tuple) and Ellipsis in index):
-            index = convert_ellipsis_to_idx(index, self.batch_size)
-        if isinstance(index, (list, range)):
-            index = torch.tensor(index, device=self.device)
-        if isinstance(index, tuple) and any(
-            isinstance(sub_index, (list, range)) for sub_index in index
+
+        if not isinstance(
+            value,
+            (TensorDictBase, dict, numbers.Number, Tensor, MemmapTensor),
         ):
-            index = tuple(
-                torch.tensor(sub_index, device=self.device)
-                if isinstance(sub_index, (list, range))
-                else sub_index
-                for sub_index in index
-            )
-        if isinstance(index, tuple) and sum(
-            isinstance(_index, str) for _index in index
-        ) not in [len(index), 0]:
-            raise IndexError(_STR_MIXED_INDEX_ERROR)
+            try:
+                # try to cast value to tensor:
+                value = self._convert_to_tensor(value)
```

Comment:
we had a long discussion at the time with @zeenolife [here](https://github.com/pytorch/rl/pull/438) IIRC and we decided that it was better to leave `set` do the convertion rather than having `__setitem__` do it. 
What is the reason to move this to `__setitiem__`?

Reply:
didn't know about this discussion. I Placed here as a mean to check for compatible type, "fail fast" if the value type is not correct and to provide a meaningful error message to the user.
Also, as mentioned in chat, I would give another shot at that discussion and think about refactoring the way preprocessing/input validation is handled. 
I personally think that, for the sake of clarity and ease of maintenance, that it would be better to place "processing"/input validation calls only where needed and constraints checked where they are "declared" instead of bury them in low level calls and resort to flags to determine whether they are applied or not.

Reply:
> I think I see your point. Aren't we at risk of missing calls to `_convert_to_tensor` by doing this? 

if we miss calls and that goes undetected it means our tests are lacking and, thus, I would consider it is a false problem. 

> Why isn't it the case already that preprocessing of input tensors is achieved only where needed? 
The other risk is that we could be calling preprocessing multiple times [...] allows us to code the key type check only once.

My sentence was probably not that clear. The main point it is about clarity and "ownership" of the checks. E.g., Why are delegating the "value" validation but not that of the keys?
  
I think the risk of multiple validation could apply here as well if you add a check not knowing/forgetting that its checked down in the chain or setting the "do_check" flags to True.

> Regarding flags, they can be changed / removed, they're private and we can do what we want with them. If there's a better alternative we can consider it.

I would decouple the input validation from the private set function into two calls:
- validation/preprocess 
- actual set

this would remove the need for the flags (*), make the check/validation explicit and "visible" at high level. If you adopt the convention that private methods do not check the inputs (**) you should have not "multiple calls".

* unless you would like to expose them to the user.
**that is you need to check only user inputs not data you produce

Reply:
> we had a long discussion at the time with @zeenolife https://github.com/pytorch/rl/pull/438 IIRC and we decided that it was better to leave set do the convertion rather than having __setitem__ do it.

@vmoens AFAIR, the idea was exactly not to risk missing `_convert_to_tensor` method, as you've already said above.

Reply:
> this would remove the need for the flags (*), make the check/validation explicit and "visible" at high level. If you adopt the convention that private methods do not check the inputs (**) you should have not "multiple calls".

Sure, that sounds reasonable. I guess something like
```python
def set(self, key, value, inplace=False):
    key = self._validate_key(key)
    value = self._validate_value(value)
    return self._set(key, value, inplace=inplace)
```
Then if we want to skip the validation we just call the private `_set` method, is that what you had in mind?

Reply:
yep, exactly. In practice it could still require some ifs or multiple validation steps but this is the idea